### PR TITLE
[SPARK-47737][PYTHON] Bump PyArrow to 10.0.0

### PR DESCRIPTION
**Pull request by itholic**
### What changes were proposed in this pull request?

This PR proposes to bump PyArrow version up to 10.0.0

### Why are the changes needed?

To leverage the new features from the latest version.

### Does this PR introduce _any_ user-facing change?

No API changes, but the PyArrow version from user-facing documentation will be changed.

### How was this patch tested?

The existing CI should pass.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

**Comment by itholic** [comment]
cc @ueshin @HyukjinKwon FYI

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
Well, it looks risky to me because Apache Arrow 15.0.2 doesn't have enough time for validation. If there is a bug in Apache Arrow 15.0.2, Apache Spark 4.0.0 will lose all PyArrow-based feature because we enforce it as the minimum.
```
- _minimum_pyarrow_version = "4.0.0"
+ _minimum_pyarrow_version = "15.0.2"
```

Are you sure about enforcing the minimum to 15.0.2, @itholic , @HyukjinKwon , @cloud-fan ?

File: python/packaging/classic/setup.py [hunk 1]
```diff
@@ -152,7 +152,7 @@ def _supports_symlinks():
 # python/packaging/connect/setup.py
 _minimum_pandas_version = "1.4.4"
 _minimum_numpy_version = "1.21"
-_minimum_pyarrow_version = "4.0.0"
+_minimum_pyarrow_version = "15.0.2"
```
Comment (HyukjinKwon) [inline python/packaging/classic/setup.py:155]: yeah doing it to the latest is too risky. Can you check the release note, and identify which version we should choose?

**Comment by itholic** [comment]
Sounds reasonable to me. We need at least PyArrow 10.0.0. Let me bump up to 10.0.0.

**Comment by itholic** [comment]
Arranged the version to 10.0.0 to minimize the potential side effect and also updated the PR title and description accordingly.

File: dev/create-release/spark-rm/Dockerfile [hunk 2]
```diff
@@ -37,7 +37,7 @@ ENV DEBCONF_NONINTERACTIVE_SEEN true
 # These arguments are just for reuse and not really meant to be customized.
 ARG APT_INSTALL="apt-get install --no-install-recommends -y"
 
-ARG PIP_PKGS="sphinx==4.5.0 mkdocs==1.1.2 numpy==1.20.3 pydata_sphinx_theme==0.13.3 ipython==7.19.0 nbsphinx==0.8.0 numpydoc==1.1.0 jinja2==3.1.2 twine==3.4.1 sphinx-plotly-directive==0.1.3 sphinx-copybutton==0.5.2 pandas==1.5.3 pyarrow==3.0.0 plotly==5.4.0 markupsafe==2.0.1 docutils<0.17 grpcio==1.62.0 protobuf==4.21.6 grpcio-status==1.62.0 googleapis-common-protos==1.56.4"
+ARG PIP_PKGS="sphinx==4.5.0 mkdocs==1.1.2 numpy==1.20.3 pydata_sphinx_theme==0.13.3 ipython==7.19.0 nbsphinx==0.8.0 numpydoc==1.1.0 jinja2==3.1.2 twine==3.4.1 sphinx-plotly-directive==0.1.3 sphinx-copybutton==0.5.2 pandas==1.5.3 pyarrow==10.0.0 plotly==5.4.0 markupsafe==2.0.1 docutils<0.17 grpcio==1.62.0 protobuf==4.21.6 grpcio-status==1.62.0 googleapis-common-protos==1.56.4"
```
Comment (HyukjinKwon) [inline dev/create-release/spark-rm/Dockerfile:40]: Shall we install highest maintenance version of 10.x?

**Review by HyukjinKwon — APPROVED** [review]
Seems fine otherwise

File: dev/create-release/spark-rm/Dockerfile (hunk 2 above)
Comment (itholic) [reply dev/create-release/spark-rm/Dockerfile]: Yeah, we can use 10.0.1 which is the latest maintenance version of 10.x. Just updated.

File: python/docs/source/getting_started/install.rst [hunk 3]
```diff
@@ -157,7 +157,7 @@ Package                    Supported version Note
 ========================== ========================= ======================================================================================
 `py4j`                     >=0.10.9.7                Required
 `pandas`                   >=1.4.4                   Required for pandas API on Spark and Spark Connect; Optional for Spark SQL
-`pyarrow`                  >=4.0.0                   Required for pandas API on Spark and Spark Connect; Optional for Spark SQL
+`pyarrow`                  >=10.0.1                  Required for pandas API on Spark and Spark Connect; Optional for Spark SQL
```
Comment (HyukjinKwon) [inline python/docs/source/getting_started/install.rst:160]: Can we don't change this upperbound though?

Comment (HyukjinKwon) [reply python/docs/source/getting_started/install.rst:160]: I am asking to keep `10.0.1` only at [dev/create-release/spark-rm/Dockerfile](https://github.com/apache/spark/pull/45892/files#diff-74d65c5e02efda4c0b26cb1d6f8b8dd08300546e5ceac4253d1428475ec206f6)

Comment (itholic) [reply python/docs/source/getting_started/install.rst:160]: Ah, I see

Comment (itholic) [reply python/docs/source/getting_started/install.rst]: Updated!

**Review by HyukjinKwon — APPROVED** [review]

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you, @itholic and @HyukjinKwon .
Merged to master.

**Comment by itholic** [comment]
Thanks for the review @dongjoon-hyun and @HyukjinKwon !

**Pull request closed without merging**